pkg/common/flyte_url.go

```diff
@@ -0,0 +1,91 @@
+package common
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/admin"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/core"
+)
+
+// transform to snake case to make lower case
+//go:generate enumer --type=ArtifactType --trimprefix=ArtifactType -transform=snake
+
+type ArtifactType int
+
+// The suffixes in these constants are used to match against the tail end of the flyte url, to keep tne flyte url simpler
+const (
+	ArtifactTypeUndefined ArtifactType = iota
+	ArtifactTypeI                      // inputs
+	ArtifactTypeO                      // outputs
+	ArtifactTypeD                      // deck
+)
+
+var re = regexp.MustCompile("flyte://v1/([a-zA-Z0-9_-]+)/([a-zA-Z0-9_-]+)/([a-zA-Z0-9_-]+)/([a-zA-Z0-9_-]+)(?:/([0-9]+))?/([iod])$")
+
+func ParseFlyteURL(flyteURL string) (core.NodeExecutionIdentifier, *int, ArtifactType, error) {
```

Comment:
I kinda like to rewrite this to embed the names of the groups in the regex itself and use this utility function to return a map of group name to value... just to encapsulate the matching/ordering logic in one place.

```suggestion

func MatchRegex(reg *regexp.Regexp, input string) map[string]string {
	names := reg.SubexpNames()
	res := reg.FindAllStringSubmatch(input, -1)

	if len(res) == 0 {
		return nil
	}

	dict := make(map[string]string, len(names))
	for i := 1; i < len(res[0]); i++ {
		dict[names[i]] = res[0][i]
	}

	return dict
}

func ParseFlyteURL(flyteURL string) (core.NodeExecutionIdentifier, *int, ArtifactType, error) {
```